console/src/hooks/useUndoableDispatch.ts

```diff
@@ -0,0 +1,81 @@
+// Copyright 2024 Synnax Labs, Inc.
+//
+// Use of this software is governed by the Business Source License included in the file
+// licenses/BSL.txt.
+//
+// As of the Change Date specified in that file, in accordance with the Business Source
+// License, use of this software will be governed by the Apache License, Version 2.0,
+// included in the file licenses/APL.txt.
+
+import { type Dispatch, type Selector, type UnknownAction } from "@reduxjs/toolkit";
+import { deep } from "@synnaxlabs/x";
+import { useCallback, useRef } from "react";
+import { useStore } from "react-redux";
+
+interface History<T> {
+  past: T[];
+  present: T | null;
+  future: T[];
+}
+
+export const useUndoableDispatch = <
+  StoreType extends {} = {},
+  SelectedType = StoreType,
+>(
+  selector: Selector<StoreType, SelectedType>,
+  actionCreator: (state: SelectedType) => UnknownAction,
+  waitFor: number = 0,
+) => {
+  const store = useStore<StoreType>();
+
+  const history = useRef<History<SelectedType>>({
+    past: [],
+    present: null,
+    future: [],
+  });
+
+  const updateState = useCallback(() => {
+    const currentState = selector(store.getState());
+    if (deep.equal(history.current.present, currentState)) return;
+    if (history.current.present != null)
+      history.current.past.push(history.current.present);
+    history.current.present = currentState;
+    history.current.future = [];
+  }, [store, selector]);
+
+  let timeoutID: NodeJS.Timeout;
```

Comment:
This variable will get re-initialized every time the hook runs, which can be independent of the `timeoutID` assignment inside of the callback. This means that this value may get set to undefined even if there's an active timeout. Use a ref here instead of a static variable.